rosidl_typesupport_cpp/QUALITY_DECLARATION.md

```diff
@@ -106,17 +113,20 @@ There are currently no tests for the public API.
 
 `rosidl_typesupport_cpp` uses and passes all the standard linters and static analysis tools for a C++ package as described in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#linters).
 
-The latest linting results can be found [here](http://build.ros2.org/view/Epr/job/Epr__rosidl_typesupport__ubuntu_bionic_amd64/lastBuild/testReport/rosidl_typesupport_cpp/).
+Results of the linting tests can be found [here](https://ci.ros2.org/job/nightly_linux_release/lastBuild/testReport/rosidl_typesupport_cpp/).
 
 ## Dependencies [5]
 
 ### Direct Runtime ROS Dependencies [5.i/5.ii]
+
 `rosidl_typesupport_cpp` has the following runtime ROS dependencies:
 * `rcpputils`
```

Comment:
Use this format https://github.com/ros2/rclcpp/blob/master/rclcpp/QUALITY_DECLARATION.md#direct-and-optional-runtime-ros-dependencies-5i5ii

Reply:
Hm.  I'm honestly don't think that format is a good idea.  It means that when we go to change the quality level, not only do we have to change it in the source repository for the package itself, we also have to change it in all of the dependent repositories as well.  That seems like a lot of churn to me.

Reply:
I agree, and we've talked about the maintenance burden before. I expected quality level automation to take care of it. @wjwwood ?

Reply:
I doubt the automation will address dependencies anytime soon due to the complexity and the timelines I think that will be involved. I wouldn't rely on that for now.

Keeping this information up to date is tedious, but also necessary if you want to stay up to date.

Reply:
> Keeping this information up to date is tedious, but also necessary if you want to stay up to date.

But I don't actually agree that it is necessary.  If you just put the quality levels in the packages they pertain to, then you only have to update it in one spot when you change the quality level.  Having it in all of the downstream dependencies mean you then have to make a PR for the package it belongs to and for all of the downstream dependencies.  That seems more than tedious, and I'm not convinced it adds a lot of benefit anyway.

Reply:
But when you're talking about whether or not your package is level 2 or 3, you need to not only point to your dependencies, but specific versions of them where they were at a similar level to your package's claim. It's not always master to master or something. And quality levels of your dependencies will change over time. The benefit of needing to update it in various places is that you let people know things are changing. An automated solution would be best, but as I said, that's some time out.

Reply:
> But when you're talking about whether or not your package is level 2 or 3, you need to not only point to your dependencies, but specific versions of them where they were at a similar level to your package's claim. It's not always master to master or something

I totally agree with this, but if you are really serious about this, I think you aren't going to take the word of the downstream package anyway.  You'd have an automated script that would walk the dependency tree looking at the QUALITY_DECLARATIONS for the entire tree.

And I'll point out that even with us actively looking at this stuff now, we already have things wrong in the QUALITY_DECLARATION.md files.  This is not to place blame on anyone, but the fact is with 54 QD files, and more to come over time, it's not going to be sustainable to have to update dozens of them when something changes.

Reply:
> I totally agree with this, but if you are really serious about this, I think you aren't going to take the word of the downstream package anyway.

To be clearer, I think we should separate the users here. For the maintainers of the package I think it is totally reasonable to say "I use this package at this version and they claim to be level X, and that's good enough for me".

For consumers, they may trust this, or rely on a third party verifying it, or check them recursively themselves, but that's already the case (whether or not you include the quality levels of direct dependencies or not). The automation would certainly help them, but they're not the primary user of having the quality levels of dependencies in a quality declaration (in my opinion). I think that's more for the benefit of casual consumers and as a way for maintainers to "show their work" that they actually looked at the dependencies at some point, if not recursively.

Either way we go, I think that this statement:

- "I use rclcpp at version x.y.z and as of MM/YYYY it was declared to be quality level 2"

Is just way more useful than:

- "I use rclcpp"

I understand the frustration, but there's just no way (that I see) getting around the maintenance burden of this kind of work. Automation would help but it is expensive too.

To me it's not an issue with the approach, its the chronic issue of not having enough time to polish packages, because we have so much software and so few maintainers. This kind of documentation and auditing is the height of project polish, and it's just going to take TLC to keep it up to date.

Reply:
So I went ahead and prototyped a solution in python to find quality declarations of dependencies: https://gist.github.com/clalancette/d0a77b0433670803cf69de73c1f05cd0 .

It's not perfect; it has at least the following flaws:
1.  There's no option to recurse into sub-dependencies.
1.  It currently only looks at `<depends>` tags.
1.  If the QUALITY_DECLARATION.md is at the repository level, rather than the package level, it will fail to find it.  This seems more like a bug in those individual packages, but I could make the script smarter as well.
1.  The final output is just a raw print of a dictionary.

Regardless, none of the above problems are show-stoppers, and with a few hours of work I could easily fix them.  But this is the direction I'd much rather see us head in, rather than duplicating all of the quality information everywhere.

Reply:
The thing that I think will take time is getting the right version for each. In the future when the foxy version is QL 3, Rolling is QL 2, and master is QL 1, how will you know which to check? This will require the user specifying which but then it will require using rosdistro. It isn't impossible, but it is some more work.

Also, what if not all the dependencies should be considered for QL? My thinking is that you need to annotate the package.xml, then have your script avoid looking at those. We need to come up with a way to do the annotation and then detect it, which might be an issue since looking for XML comments or something may not be trivial.

Reply:
> The thing that I think will take time is getting the right version for each. In the future when the foxy version is QL 3, Rolling is QL 2, and master is QL 1, how will you know which to check?

In the short-term, the easiest way I can think of to do that would be to clone https://github.com/ros2/ros2/blob/master/ros2.repos at the revision you are interested in looking at, then running the script.  If you want to look at packages outside of the ROS 2 core, then you could do the same sort of thing with `rosinstall_generator` at the revision you are interested in.

> Also, what if not all the dependencies should be considered for QL?

It's an interesting point.  Could you give an example of what you are thinking of here?

I can think of 2 different ways to handle it; either on the producer (package.xml) or consumer (script) side.  On the producer side, you would have to do what you said; add some annotation to package.xml, and then respect that.  On the consumer side, we could augment the script to take a list of exceptions to ignore.  The benefit of the latter is that we don't need to iterate package.xml, and it seems like that exclusion would be a decision the consumer would want to make (not the producer).  But it sort of depends on the situations you are thinking of.

Reply:
I ended up spending some more time on this, and ended up fixing most of the issues that were pointed out above.  The tool is now at https://github.com/clalancette/walk-qds , and can walk dependencies recursively.  It now looks at both `<depend>` and `<build_depend>` tags (though it could be trivially extended to look at more).  I've also added a command-line option to exclude certain dependencies from being considered.  It also fixes a bunch of bugs in my earlier prototype.

I'd encourage people to give it a try and see if it fits what we are looking to do here.  If it does, I'd like to try to move forward with this PR as-is, and then go back and remove the (what I consider) redundant information from rclcpp.

Reply:
I'm not here to weigh in on documenting everything vs automation. But I can suggest one possible approach for documentation that I believe should result in minimal maintenance, unless an upstream package downgrades their status.

State that each dependency is at least "QL 3", then link to the location where their current quality levels can be found.
For example: https://github.com/ros2/rcl_interfaces/blob/master/action_msgs/QUALITY_DECLARATION.md#direct-runtime-ros-dependencies-5i5ii